WalletWasabi.Gui/Rpc/JsonRpcServer.cs

```diff
@@ -0,0 +1,108 @@
+using System;
+using System.IO;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Gui.Models;
+
+namespace WalletWasabi.Gui.Rpc
+{
+    public class JsonRpcServer
+    {
+        private long _running;
+        public bool IsRunning => Interlocked.Read(ref _running) == 1;
+        public bool IsStopping => Interlocked.Read(ref _running) == 2;
+        private CancellationTokenSource Cancellation { get; }
+
+        private HttpListener Listener { get; }
+        private WasabiJsonRpcService Service { get; }
+        private JsonRpcServerConfiguration Config { get; }
+
+        public JsonRpcServer(Global global, JsonRpcServerConfiguration config)
+        {
+            Config = config;
+            Listener = new HttpListener();
+            Listener.AuthenticationSchemes = AuthenticationSchemes.Basic | AuthenticationSchemes.Anonymous;
+            foreach (var prefix in Config.Prefixes)
+            {
+                Listener.Prefixes.Add(prefix);
+            }
+            Cancellation = new CancellationTokenSource();
+            Service = new WasabiJsonRpcService(global);
+        }
+
+        public void Start()
+        {
+            Interlocked.Exchange(ref _running, 1);
+            Listener.Start();
+
+            Task.Factory.StartNew(async _ =>
```

Comment:
I think you can use `PeriodicRunner` here.

Reply:
I thought that too but this code doesn't run periodically.